File: domain/src/main/java/org/openmedstack/domain/SagaBase.java
package org.openmedstack.domain;

import openmedstack.commands.DomainCommand;
import openmedstack.events.BaseEvent;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.function.Consumer;

public abstract class SagaBase implements Saga {
    private final IRouteEvents _eventRouter;
    private final ArrayList<BaseEvent> _uncommitted = new ArrayList<BaseEvent>();
    private final ArrayList<DomainCommand> _undispatched = new ArrayList<DomainCommand>();
    private final String _id;
    private int _version = 0;

    protected SagaBase(String id, IRouteEvents eventRouter) {
        _eventRouter = eventRouter == null ? new ConventionEventRouter(true, this) : eventRouter;
        _id = id;
    }

    @Override
    public String getId() {
        return _id;
    }

    @Override
    public Integer getVersion() {
        return _version;
    }

    @Override
    public void transition(Object message) throws HandlerForDomainEventNotFoundException {
        if (!BaseEvent.class.isInstance(message)) {
            return;
        }

        _eventRouter.dispatch(message);

        _uncommitted.add((BaseEvent) message);
        _version += 1;
    }

    public Iterable<Object> getUncommittedEvents() {
        return Arrays.asList(_uncommitted);
    }

    public void clearUncommittedEvents() {
        _uncommitted.clear();
    }

    public Iterable<Object> getUndispatchedMessages() {
        return Arrays.asList(_undispatched);
    }

    public void clearUndispatchedMessages() {
        _undispatched.clear();
    }

    protected void dispatch(DomainCommand message) {
        _undispatched.add(message);
    }

    protected void register(Consumer<BaseEvent> handler) {
        _eventRouter.register(handler);
    }
}

File: domain/src/main/java/org/openmedstack/domain/SagaBase.kt
package org.openmedstack.domain

import org.openmedstack.commands.DomainCommand
import org.openmedstack.events.BaseEvent
import java.util.*
import java.util.function.Consumer

abstract class SagaBase protected constructor(id: String, eventRouter: IRouteEvents?) : Saga {
    private val _eventRouter: IRouteEvents
    private val _uncommitted: ArrayList<BaseEvent> = ArrayList()
    private val _undispatched: ArrayList<DomainCommand> = ArrayList()
    private val _id: String
    private var _version = 0

    override val id: String
        get() = _id

    override val version: Int
        get() = _version

    @Throws(HandlerForDomainEventNotFoundException::class)
    override fun transition(message: Any) {
        if (!BaseEvent::class.java.isInstance(message)) {
            return
        }
        _eventRouter.dispatch(message)
        _uncommitted.add(message as BaseEvent)
        _version += 1
    }

    override val uncommittedEvents: Iterable<Any>
        get() =  listOf<Any>(_uncommitted)

    override fun clearUncommittedEvents() {
        _uncommitted.clear()
    }

    override val undispatchedMessages: Iterable<Any>
        get() = listOf<Any>(_undispatched)

    override fun clearUndispatchedMessages() {
        _undispatched.clear()
    }

    protected fun dispatch(message: DomainCommand) {
        _undispatched.add(message)
    }

    protected fun register(handler: Consumer<BaseEvent>) {
        _eventRouter.register(handler)
    }

    init {
        _eventRouter = eventRouter ?: ConventionEventRouter(true, this)
        _id = id
    }
}
